# EDA API

### Importación de librerías

In [1]:
import sys
import os
import requests
import pandas as pd
import numpy as np
from datetime import datetime, time
import seaborn as sns
import matplotlib.pyplot as plt
import math
sys.path.append(os.path.abspath('../source'))


### Descarga de Datos de Accidentes FARS desde la API de NHTSA

In [2]:
'''
base_url = "https://crashviewer.nhtsa.dot.gov/CrashAPI/FARSData/GetFARSData"

# Encabezados
headers = {
    "Accept": "text/csv", 
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"
}


output_dir = "../data"
os.makedirs(output_dir, exist_ok=True)


for year in range(2017, 2023):  

    url = f"{base_url}?dataset=Accident&FromYear={year}&ToYear={year}&State=*&format=csv"
    print(f"Descargando datos para el año {year}...")

    try:
        # Realizar la solicitud al API
        response = requests.get(url, headers=headers, timeout=600, stream=True)  # Timeout de 10 minutos

        # Verificar si la solicitud fue exitosa
        if response.status_code == 200:
            # Guardar los datos en el archivo CSV
            output_file = os.path.join(output_dir, f"FARS_data_{year}.csv")
            with open(output_file, "wb") as file:
                for chunk in response.iter_content(chunk_size=1024):
                    if chunk:
                        file.write(chunk)
            print(f"Datos del año {year} guardados exitosamente en {output_file}")
        else:
            print(f"Error al obtener los datos para el año {year}: Código HTTP {response.status_code}")
            print(response.text)

    except requests.exceptions.Timeout:
        print(f"La solicitud para el año {year} excedió el tiempo límite.")
    except requests.exceptions.RequestException as e:
        print(f"Error en la solicitud para el año {year}: {e}")
'''

'\nbase_url = "https://crashviewer.nhtsa.dot.gov/CrashAPI/FARSData/GetFARSData"\n\n# Encabezados\nheaders = {\n    "Accept": "text/csv", \n    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"\n}\n\n\noutput_dir = "../data"\nos.makedirs(output_dir, exist_ok=True)\n\n\nfor year in range(2017, 2023):  \n\n    url = f"{base_url}?dataset=Accident&FromYear={year}&ToYear={year}&State=*&format=csv"\n    print(f"Descargando datos para el año {year}...")\n\n    try:\n        # Realizar la solicitud al API\n        response = requests.get(url, headers=headers, timeout=600, stream=True)  # Timeout de 10 minutos\n\n        # Verificar si la solicitud fue exitosa\n        if response.status_code == 200:\n            # Guardar los datos en el archivo CSV\n            output_file = os.path.join(output_dir, f"FARS_data_{year}.csv")\n            with open(output_file, "wb") as file:\n                for chunk in response.iter_content(chunk_size=1024):\n                    if chunk:\n           

In [3]:
'''
base_url = "https://crashviewer.nhtsa.dot.gov/CrashAPI/FARSData/GetFARSData"

# Encabezados
headers = {
    "Accept": "text/csv", 
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"
}

output_dir = "../data"
os.makedirs(output_dir, exist_ok=True)

for year in range(2017, 2023):  
    
    url = f"{base_url}?dataset=Person&FromYear={year}&ToYear={year}&State=*&format=csv"
    print(f"Descargando datos para el año {year}...")

    try:
        # Realizar la solicitud al API
        response = requests.get(url, headers=headers, timeout=600, stream=True)  # Timeout de 10 minutos

        # Verificar si la solicitud fue exitosa
        if response.status_code == 200:
            # Guardar los datos en el archivo CSV
            output_file = os.path.join(output_dir, f"FARS_person_{year}.csv")
            with open(output_file, "wb") as file:
                for chunk in response.iter_content(chunk_size=1024):
                    if chunk:
                        file.write(chunk)
            print(f"Datos del año {year} guardados exitosamente en {output_file}")
        else:
            print(f"Error al obtener los datos para el año {year}: Código HTTP {response.status_code}")
            print(response.text)

    except requests.exceptions.Timeout:
        print(f"La solicitud para el año {year} excedió el tiempo límite.")
    except requests.exceptions.RequestException as e:
        print(f"Error en la solicitud para el año {year}: {e}")
'''

'\nbase_url = "https://crashviewer.nhtsa.dot.gov/CrashAPI/FARSData/GetFARSData"\n\n# Encabezados\nheaders = {\n    "Accept": "text/csv", \n    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"\n}\n\noutput_dir = "../data"\nos.makedirs(output_dir, exist_ok=True)\n\nfor year in range(2017, 2023):  \n\n    url = f"{base_url}?dataset=Person&FromYear={year}&ToYear={year}&State=*&format=csv"\n    print(f"Descargando datos para el año {year}...")\n\n    try:\n        # Realizar la solicitud al API\n        response = requests.get(url, headers=headers, timeout=600, stream=True)  # Timeout de 10 minutos\n\n        # Verificar si la solicitud fue exitosa\n        if response.status_code == 200:\n            # Guardar los datos en el archivo CSV\n            output_file = os.path.join(output_dir, f"FARS_person_{year}.csv")\n            with open(output_file, "wb") as file:\n                for chunk in response.iter_content(chunk_size=1024):\n                    if chunk:\n               

#### Carga de Datos de Accidentes FARS desde Archivos CSV

In [4]:
# Definir los años a procesar
years = range(2017, 2023)  # Desde 2017 hasta 2022

# Lista para almacenar DataFrames
dataframes = []

# Cargar archivos en un ciclo
for year in years:
    file_path = f'../data/FARS_data_{year}.csv'  # Ajusta el path si es necesario
    df = pd.read_csv(file_path)
    df["Year"] = year  # Agregar la columna de año para referencia
    dataframes.append(df)

# Concatenar todos los DataFrames en uno solo
accidents = pd.concat(dataframes, ignore_index=True)

# Ver los primeros registros
accidents.head()

/tmp/ipykernel_58825/1353280254.py:10: DtypeWarning: Columns (41,43) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


caseyear  state  st_case statename  ve_total  ve_forms  pvh_invl  peds  \
0      2017      1    10001   Alabama         1         1         0     0   
1      2017      1    10002   Alabama         1         1         0     0   
2      2017      1    10003   Alabama         3         3         0     0   
3      2017      1    10004   Alabama         1         1         0     0   
4      2017      1    10005   Alabama         1         1         0     0   

   pernotmvit  permvit  ...                       hosp_mnname   cf1  \
0           0        1  ...  Not Applicable (Not Transported)   0.0   
1           0        1  ...  Not Applicable (Not Transported)   0.0   
2           0        3  ...  Not Applicable (Not Transported)   0.0   
3           0        1  ...  Not Applicable (Not Transported)  20.0   
4           0        2  ...                                11   0.0   

                   cf1name  cf2 cf2name  cf3  cf3name  fatals drunk_dr  Year  
0                      NaN  0.0     NaN  0.0      NaN       1      0.0  2017  
1                      NaN  0.0     NaN  0.0      NaN       1      0.0  2017  
2                      NaN  0.0     NaN  0.0      NaN       1      0.0  2017  
3  Police Pursuit Involved  0.0     NaN  0.0      NaN       1      0.0  2017  
4                      NaN  0.0     NaN  0.0      NaN       1      0.0  2017  

[5 rows x 95 columns]

Esta celda carga los datos delos accidentes desde archivos CSV para los años 2017 a 2022

In [5]:
# Años a procesar
years = range(2017, 2023)  # Hasta 2022 

# Lista para almacenar DataFrames
dataframes = []

# Cargar archivos en un ciclo
for year in years:
    file_path = f'../data/FARS_person_{year}.csv'  # Ajusta el path si es necesario
    df = pd.read_csv(file_path)
    df["Year"] = year  # Agregar la columna de año para referencia
    dataframes.append(df)

# Concatenar todos los DataFrames en uno solo
persons = pd.concat(dataframes, ignore_index=True)

/tmp/ipykernel_58825/517759701.py:10: DtypeWarning: Columns (117,119) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)
/tmp/ipykernel_58825/517759701.py:10: DtypeWarning: Columns (18,117,119) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)
/tmp/ipykernel_58825/517759701.py:10: DtypeWarning: Columns (18,115,117,119) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)
/tmp/ipykernel_58825/517759701.py:10: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)
/tmp/ipykernel_58825/517759701.py:10: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)
/tmp/ipykernel_58825/517759701.py:10: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memor

Esta celda carga los datos de personas involucradas en los accidentes desde archivos CSV para los años 2017 a 2022:

#### Unión de Datos de Accidentes y Persons 

In [6]:
print(accidents.columns)  # Muestra todas las columnas disponibles

Index(['caseyear', 'state', 'st_case', 'statename', 've_total', 've_forms',
       'pvh_invl', 'peds', 'pernotmvit', 'permvit', 'persons', 'county',
       'countyname', 'city', 'cityname', 'day', 'dayname', 'month',
       'monthname', 'year', 'day_week', 'day_weekname', 'hour', 'hourname',
       'minute', 'minutename', 'nhs', 'nhsname', 'rur_urb', 'rur_urbname',
       'func_sys', 'func_sysname', 'rd_owner', 'rd_ownername', 'route',
       'routename', 'tway_id', 'tway_id2', 'milept', 'mileptname', 'latitude',
       'latitudename', 'longitud', 'longitudname', 'sp_jur', 'sp_jurname',
       'harm_ev', 'harm_evname', 'man_coll', 'man_collname', 'reljct1',
       'reljct1name', 'reljct2', 'reljct2name', 'typ_int', 'typ_intname',
       'wrk_zone', 'wrk_zonename', 'road_fnc', 'road_fncname', 'rel_road',
       'rel_roadname', 'lgt_cond', 'lgt_condname', 'weather1', 'weather1name',
       'weather2', 'weather2name', 'weather', 'weathername', 'sch_bus',
       'sch_busname', 'rail', 'rai

imprime los nombres de las columnas del DataFrame accidents para inspeccionar las variables disponibles.

La lista de columnas proporciona una visión completa de las variables disponibles en los datos de accidentes, lo que facilita la selección de columnas relevantes para la unión con los datos de personas y el análisis posterior.

In [ ]:
print(persons.columns)  

Index(['caseyear', 'state', 'statename', 'st_case', 've_forms', 'veh_no',
       'per_no', 'str_veh', 'str_vehname', 'county',
       ...
       'icfinalbodyname', 'gvwr_from', 'gvwr_fromname', 'gvwr_to',
       'gvwr_toname', 'devtype', 'devtypename', 'devmotor', 'devmotorname',
       'Year'],
      dtype='object', length=160)


: 

imprime los nombres de las columnas del DataFrame persons para inspeccionar las variables disponibles.

La gran cantidad de columnas (160) indica que los datos de personas son muy detallados, con información sobre cada individuo involucrado en los accidentes. Columnas como age, sex, y alc_res son clave para el análisis demográfico y de comportamiento.

In [ ]:
df = accidents.merge(persons[['st_case', 'age', 'sex', 'alc_res']], on='st_case', how='left')

Esta celda combina los DataFrames accidents y persons usando un *merge* por la columna st_case (identificador único del accidente):

- Selecciona solo las columnas st_case, age, sex, y alc_res del DataFrame persons.
- Usa un *left join* para mantener todos los registros de accidents y agregar la información de personas correspondiente, dejando valores nulos si no hay coincidencia.

La unión permite combinar información de accidentes (como condiciones del accidente y fatalidades) con datos demográficos y de alcohol de las personas involucradas, preparando el conjunto de datos para un análisis más detallado.

In [ ]:
df_copy = df.copy()

In [ ]:
df.head()


### Generalidades

In [ ]:
df.shape



Esta celda muestra el tamaño del DataFrame `df` (3,012,888 filas, 98 columnas)

El DataFrame contiene una gran cantidad de datos (más de 3 millones de registros), con un registro por cada persona involucrada en un accidente. Las 98 columnas indican una amplia variedad de variables para analizar.

#### Tipos de Datos

In [ ]:
df.info()


El DataFrame contiene una amplia variedad de datos (numéricos, categóricos y de texto) debido a la mezcla de tipos de datos.

- **Tamaño**: 3,012,888 filas y 98 columnas.  
- **Tipos de datos**: 47 columnas `int64`, 10 columnas `float64`, 41 columnas `object`.  
- **Columnas clave**: `caseyear` (int64), `statename` (object), `latitude` (float64), `age` (int64), `alc_res` (int64).  
- **Uso de memoria**: Más de 2.2 GB, lo que indica un conjunto de datos grande.


In [ ]:
df.columns

### Limpieza

In [ ]:
columns = [
    've_total', 'fatals', 'peds','arr_hour', 'arr_min',
    'year', 'monthname', 'day_weekname','hour','minute',
    'state', 'statename', 'rur_urbname', 'func_sysname', 
    'weathername', 'lgt_condname','harm_evname',
    'age', 'sex', 'alc_res'
]

df = df[columns]


In [ ]:
df.head()

In [ ]:
df.dtypes

#### Manejo de datos de faltantes

In [ ]:
df.isnull().sum()

#### Manejo de datos duplicados

In [ ]:
df.duplicated().sum()

In [ ]:
df[df.duplicated()]

In [ ]:
df = df.drop_duplicates()

In [ ]:
df.shape

In [ ]:
df['rur_urbname'].value_counts()    

In [ ]:
df = df.loc[~df["rur_urbname"].isin(["Not Reported", "Trafficway Not in State Inventory", "Unknown"])].copy()

In [ ]:
df['weathername'].value_counts()  

In [ ]:
df = df.loc[~df["weathername"].isin([
    "Not Reported", "Other", "Unknown", "Reported as Unknown"
])].copy()

In [ ]:
df.shape

In [ ]:
df['lgt_condname'].value_counts()  

In [ ]:
df = df.loc[~df['lgt_condname'].isin([
    "Not Reported", "Other", "Unknown", "Reported as Unknown"
])].copy()

In [ ]:
# Diccionario con códigos especiales por columna
special_codes = {
    'arr_hour': [88, 99],
    'arr_min': [88, 97, 98, 99],
}

for col, codes in special_codes.items():
    df[col] = df[col].replace({code: -1 for code in codes})  # -1 como marcador en lugar de NaN


In [ ]:
df = df.loc[~df[['arr_hour', 'arr_min']].isin([-1]).any(axis=1)].copy()



Las siguientes variables de tiempo del dataset FARS fueron procesadas para eliminar códigos especiales o valores fuera del rango válido:

- `arr_hour`: hora de llegada
- `arr_min`: minuto de llegada


Estas variables contenían códigos utilizados por el sistema FARS para representar valores especiales:

| Código | Significado común en FARS |
|--------|----------------------------|
| `88`   | Not Applicable             |
| `96`   | Anómalo (fuera de rango)   |
| `97`   | Imputed                    |
| `98`   | Estimated                  |
| `99`   | Unknown / Missing          |

Estos códigos fueron reemplazados por `-1` y eliminados dado que no eran muchos.

Por ejemplo:
- `arr_min` contenía valores como `97`, `98`, `99`, que no representan minutos válidos.

### Manejo de la variable `sex`

In [ ]:
df["sex"].value_counts()[[8, 9]]

In [ ]:
df = df.loc[~df["sex"].isin(["Not Reported", "Unknown"])].copy()

In [ ]:
sex_mapping = {
    1: "Male",
    2: "Female",
    3: "Other",
}

df["sex"] = df["sex"].map(sex_mapping)

### Manejo de la variable `func_sysname`

In [ ]:
df['func_sysname'].unique()

In [ ]:
# Diccionario con códigos especiales para la columna 'func_sysname'
special_codes = {
    'func_sysname': ['Not Reported', 'Unknown']  
}

# Reemplazar valores especiales por -1 como marcador en lugar de NaN
for col, codes in special_codes.items():
    df[col] = df[col].replace({code: -1 for code in codes})

In [ ]:
# Filtrar filas donde 'func_sysname' no contenga valores especiales (-1)
df = df.loc[~df['func_sysname'].isin([-1])].copy()

In [ ]:
df['func_sysname'].unique()

## Estadísticas descriptivas

In [ ]:
df.describe(include="object")

In [ ]:
df.describe().T

Se presentan estadísticas para columnas numéricas como `ve_total`, `fatals`, `peds`, `arr_hour`, `arr_min`, `year`, `hour`, `minute`, `state`, `age`, `alc_res`:  
- **`ve_total` (número de vehículos)**: Media de 1.66, máximo de 50. La mayoría de los accidentes involucran pocos vehículos (1-2).  
- **`fatals` (fatalidades)**: Media de 1.10, máximo de 20. La mayoría de los accidentes tienen 1 fatalidad.  
- **`peds` (peatones)**: Media de 0.20, máximo de 20. Los peatones están involucrados en pocos casos.  
- **`age` (edad)**: Media de 57.70, máximo de 999. Valores como 999 sugieren datos faltantes o "desconocido".  
- **`alc_res` (nivel de alcohol)**: Media de 632.27, máximo de 999. Muchos valores en 996 indican "sin prueba de alcohol".  

La mayoría de los accidentes involucran 1-2 vehículos y 1 fatalidad. Los valores atípicos en `age` (999) y `alc_res` (996, 999) indican datos faltantes o codificados como "desconocido", lo que requerirá limpieza. La distribución de horas (`hour`, `arr_hour`) muestra que los accidentes ocurren durante todo el día, con un promedio alrededor de las 13:00-14:00.

### Análisis de valores atípicos

In [ ]:
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns

outlier_log = []

for col in numerical_cols:
    print(f"\n--- Analyzing Outliers for '{col}' ---")

    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_limit = Q1 - 1.5 * IQR
    upper_limit = Q3 + 1.5 * IQR

    outliers = df[(df[col] < lower_limit) | (df[col] > upper_limit)]
    num_outliers = len(outliers)

    print(f"Number of Outliers: {num_outliers}")
    print(f"Lower Limit: {lower_limit:.2f}, Upper Limit: {upper_limit:.2f}")

    outlier_log.append({
        'Variable': col,
        'Outliers': num_outliers,
        'Lower Limit': lower_limit,
        'Upper Limit': upper_limit,
        'Porcentaje de Outliers': round(num_outliers / len(df) * 100, 2)
    })

print("\nResumen de Outliers:")
print(outlier_log)



Los altos porcentajes de atípicos en `fatals` y `peds` indican que los accidentes con múltiples fatalidades o peatones son raros pero relevantes. Los atípicos en `age` y `alc_res` probablemente provienen de valores codificados como "desconocido" (999, 996), lo que requiere limpieza para análisis futuros.

Se analizaron los valores atípicos en columnas como `ve_total`, `fatals`, `peds`, `arr_hour`, `arr_min`, `year`, `hour`, `minute`, `state`, `age`, `alc_res`:  
- **`ve_total` (número de vehículos)**: 37,401 atípicos (2.98%), con un límite superior de 3.5 vehículos.  
- **`fatals` (fatalidades)**: 103,454 atípicos (8.25%), con un límite superior de 1. Cualquier accidente con más de 1 fatalidad es atípico.  
- **`peds` (peatones)**: 225,024 atípicos (17.94%), con un límite superior de 0. Cualquier accidente con peatones es atípico.  
- **`age` (edad)**: 22,747 atípicos (1.81%), con un límite superior de 106.5, probablemente por valores como 999.  
- **`alc_res` (nivel de alcohol)**: 0 atípicos, pero esto puede ser engañoso debido a valores codificados como 996 o 999.  



# Transformations for the dimensional modeling

In [ ]:
df.columns


Lista de 20 columnas: ve_total, fatals, peds, arr_hour, arr_min, year, monthname, day_weekname, hour, minute, state, statename, rur_urbname, func_sysname, weathername, lgt_condname, harm_evname, age, sex, alc_res.
Interpretación:

La lista confirma las columnas disponibles para las transformaciones, incluyendo variables clave para el modelado dimensional como condiciones del accidente, características demográficas, y resultados de alcohol.

In [ ]:
df = df.drop(columns=['statename', 'state'])

In [ ]:
df["lgt_condname"].unique()

Las categorías reflejan los niveles de iluminación durante los accidentes, desde luz diurna hasta oscuridad sin iluminación.

In [ ]:
visibility_mapping = {
    'Daylight': 'Alta',
    'Dawn': 'Moderada',
    'Dusk': 'Moderada',
    'Dark - Lighted': 'Baja',
    'Dark - Not Lighted': 'Muy Baja',
    'Dark - Unknown Lighting': 'Muy Baja',
    'Unknown': 'Unknown',
    'Not Reported': 'Unknown',
    'Reported as Unknown': 'Unknown',
    'Other': 'Unknown'
}

df["lgt_condname"] = df["lgt_condname"].map(visibility_mapping)

La transformación estandariza las condiciones de iluminación en niveles de visibilidad más interpretables, facilitando el análisis en el modelo dimensional. La agrupación de categorías raras en Unknown reduce la complejidad.

In [ ]:
print(df['alc_res'].unique())

La gran cantidad de valores únicos sugiere que alc_res contiene datos continuos de BAC, junto con códigos especiales (996, 999) para datos faltantes o no reportados. Esto requiere una transformación para categorizar los valores.

In [ ]:
# Convertir el BAC a su valor real en g/dL
df['alc_res'] = df['alc_res'] / 1000

# Función de clasificación
def clasificar_bac(bac):
    if bac <= 0.03:
        return "Bajo"
    elif bac <= 0.08:
        return "Moderado"
    elif bac <= 0.20:
        return "Alto"
    elif bac <= 0.40:
        return "Peligroso"
    else:
        return "Letal"

# Aplicar la función
df['alc_res'] = df['alc_res'].apply(clasificar_bac)

# Mostrar el resultado
print(df.head())


La transformación convierte los valores numéricos de BAC en categorías interpretables, lo que facilita el análisis en el modelo dimensional. Los valores altos como 996/1000=0.996 se clasifican como Letal, lo que podría indicar la necesidad de revisar los códigos especiales (996, 999) antes de la transformación.

In [ ]:
df["weathername"].unique()

Esta celda muestra los valores únicos de la columna weathername (condiciones climáticas) para entender las categorías disponibles.

Las categorías reflejan diversas condiciones climáticas que pueden influir en los accidentes. Esto es útil para estandarizar las categorías en el modelado dimensional.

In [ ]:
weather_map = {
    'Clear': 'Clear',
    'Rain': 'Rainy',
    'Cloudy': 'Windy',
    'Fog, Smog, Smoke': 'Foggy',
    'Snow': 'Snowy',
    'Sleet or Hail': 'Snowy',
    'Freezing Rain or Drizzle': 'Rainy',
    'Blowing Snow': 'Snowy',
    'Blowing Sand, Soil, Dirt': 'Windy',
    'Severe Crosswinds': 'Windy',
    'Clear': 'Clear',
    'Not Reported': 'Unknown',
    'Other': 'Unknown',
    'Reported as Unknown': 'Unknown'
    # 'Unknown' se mantiene igual
}

# Aplicamos el reemplazo
df['weathername'] = df['weathername'].replace(weather_map)


Esta celda transforma la columna `weathername` mapeando las condiciones climáticas a categorías simplificadas:  
- `Clear` → `Clear`  
- `Rain`, `Freezing Rain or Drizzle` → `Rainy`  
- `Cloudy`, `Blowing Sand, Soil, Dirt`, `Severe Crosswinds` → `Windy`  
- `Snow`, `Sleet or Hail`, `Blowing Snow` → `Snowy`  
- `Fog`, `Smog`, `Smoke` → `Foggy`  
- Otros valores → `Unknown`

La transformación simplifica las categorías climáticas al agrupar condiciones similares, facilitando el análisis en el modelo dimensional.

In [ ]:
accidents['arr_hour'].unique()

Los valores representan horas del día (0-23), pero incluyen códigos especiales como 99 (probablemente "desconocido") y 88 (posiblemente "no reportado"). Esto indica la necesidad de limpiar estos valores antes de usarlos.

In [ ]:
df[df["arr_hour"] == 99]["arr_hour"].nunique()

In [ ]:
df[df["arr_hour"] == 88]["arr_hour"].nunique()

In [ ]:
df[df["arr_min"].isin([88, 97, 98, 99])]["arr_hour"].nunique()

In [ ]:
df['arr_min'].unique()

Los valores representan minutos (0-59) sin códigos especiales (88, 97, 98, 99), confirmando que los datos de arr_min son válidos y están listos para transformaciones de tiempo.

In [ ]:
from datetime import time

# Aseguramos que no haya NaN y convertimos
df['arr_hour'] = df['arr_hour'].fillna(0).astype(int)
df['arr_min'] = df['arr_min'].fillna(0).astype(int)

# Creamos la columna con objetos tipo datetime.time
df['arr_time'] = df.apply(lambda row: time(row['arr_hour'], row['arr_min']), axis=1)

print(df[['arr_hour', 'arr_min', 'arr_time']])


La creación de arr_time estandariza las horas de llegada en un formato de tiempo, facilitando cálculos posteriores como la diferencia de tiempo entre el accidente y la llegada de servicios de emergencia.

In [ ]:
df['minute'].unique()

In [ ]:
df['hour'].unique()

In [ ]:
df = df[~df['hour'].isin([99, 98])]
df = df[~df['minute'].isin([99, 98])]

# Aseguramos que no haya NaN y convertimos
df['hour'] = df['hour'].fillna(0).astype(int)
df['minute'] = df['minute'].fillna(0).astype(int)

# Creamos la columna con objetos tipo datetime.time
df['time'] = df.apply(lambda row: time(row['hour'], row['minute']), axis=1)

print(df[['hour', 'minute', 'time']])

La limpieza elimina datos no válidos (98, 99), y la creación de `time` estandariza las horas de los accidentes en un formato de tiempo, preparando los datos para cálculos posteriores.

In [ ]:
df = df[~df['hour'].isin([99, 98])]
df = df[~df['minute'].isin([99, 98])]


# Función para calcular la diferencia entre dos objetos time
def time_diff(t1, t2):
    df = df[~df['hour'].isin([99, 98])]
df = df[~df['minute'].isin([99, 98])]

    # Convertir a datetime con una fecha ficticia
    dt1 = datetime(2023, 1, 1, t1.hour, t1.minute, t1.second)
    dt2 = datetime(2023, 1, 1, t2.hour, t2.minute, t2.second)
    
    # Calcular diferencia (arr_time - time)
    diff = dt2 - dt1
    
    # Si arr_time es menor que time, asumir que es del día siguiente
    if diff.total_seconds() < 0:
        dt2 = datetime(2023, 1, 2, t2.hour, t2.minute, t2.second)  # Día siguiente
        diff = dt2 - dt1
    
    # Convertir la diferencia a horas, minutos, segundos
    total_seconds = diff.total_seconds()
    hours = int(total_seconds // 3600) % 24  # Tomar módulo 24 para mantener horas en 0-23
    minutes = int((total_seconds % 3600) // 60)
    seconds = int(total_seconds % 60)
    
    return time(hours, minutes, seconds)

# Manejar valores nulos y calcular la diferencia
df['diff_time'] = df.apply(
    lambda row: time_diff(row['time'], row['arr_time']) 
    if pd.notnull(row['time']) and pd.notnull(row['arr_time']) 
    else None, 
    axis=1
)

# Mostrar el resultado
print(df[['time', 'arr_time', 'diff_time']])

La columna diff_time mide el tiempo de respuesta de los servicios de emergencia, un indicador clave para el análisis de accidentes. La lógica para manejar diferencias negativas asegura que los cálculos sean correctos incluso si la llegada ocurre después de la medianoche.

In [ ]:
# Función para categorizar la hora
def categorize_time(row):
    hour = row.hour
    if 5 <= hour < 12:
        return "Morning"
    elif 12 <= hour < 17:
        return "Afternoon"
    elif 17 <= hour < 21:
        return "Evening"
    else:
        return "Night"

# Aplicamos la función a la columna time para crear la nueva columna categórica
df['time'] = df['time'].apply(categorize_time).astype("category")

print(df[['time']])

Esta celda categoriza la columna `time` (hora del accidente) en momentos del día:  
- Define una función `categorize_time` que clasifica la hora:  
  - 5:00-11:59 → `Morning`  
  - 12:00-16:59 → `Afternoon`  
  - 17:00-20:59 → `Evening`  
  - Otros (21:00-4:59) → `Night`  
- Aplica la función a `time` y convierte el resultado en una columna categórica.  
- Muestra la columna `time`.

La categorización simplifica el análisis de patrones temporales, permitiendo identificar si los accidentes son más frecuentes en ciertos momentos del día (por ejemplo, noche vs. mañana).

In [ ]:
df.columns

La lista confirma las transformaciones realizadas (nuevas columnas como arr_time, time, diff_time, time_of_day) y las columnas originales mantenidas.

In [ ]:
df['func_sysname'].unique()

Tabla de correspondencias entre `func_sysname` y los valores únicos en `'road_type'` del modelo dimensional:

|  `func_sysname` | Correspondencia con `road_type` |
|------------|--------------------------------|
| Interstate | Highway                        |
| Principal Arterial – Other Freeways and Expressways | Highway |
| Principal Arterial – Other | Main Road |
| Minor Arterial | Main Road |
| Major Collector | Main Road |
| Minor Collector | Street |
| Local | Street |


In [ ]:
# Diccionario de mapeo entre 'func_sysname' y 'road_type'
mapping = {
    'Interstate': 'Highway',
    'Principal Arterial - Other Freeways and Expressways': 'Highway',
    'Principal Arterial - Other': 'Main Road',
    'Minor Arterial': 'Main Road',
    'Major Collector': 'Main Road',
    'Minor Collector': 'Street',
    'Local': 'Street'
}

# Aplicar el mapeo en la columna 'func_sysname'
df['func_sysname'] = df['func_sysname'].map(mapping)

La transformación simplifica los tipos de carreteras en categorías más generales, facilitando el análisis en el modelo dimensional.

- `Interstate`, `Principal Arterial - Other Freeways and Expressways` → `Highway`  
- `Principal Arterial - Other`, `Minor Arterial`, `Major Collector` → `Main Road`  
- `Minor Collector`, `Local` → `Street`



In [ ]:
df['func_sysname'].unique()

La transformación fue exitosa, reduciendo las categorías de func_sysname a tres tipos de carreteras, listos para el modelado dimensional.

In [ ]:
df["age"].unique()

Los valores representan edades reales, pero la presencia de 998 y 999 indica datos faltantes que necesitan limpieza.

In [ ]:
# Elimina las filas con edades inválidas (998 y 999)
df = df[~df['age'].isin([998, 999])]

def map_driver_age(age):
    if age < 18:
        return '<18'
    elif 18 <= age <= 25:
        return '18-25'
    elif 26 <= age <= 40:
        return '26-40'
    elif 41 <= age <= 60:
        return '41-60'
    elif age > 60:
        return '61+'

df['age'] = df['age'].apply(map_driver_age)

print(df['age'])



La agrupación de edades en rangos simplifica el análisis demográfico, permitiendo identificar patrones según los grupos etarios en el modelo dimensional.
- `<18`: Menores de 18 años  
- `18-25`: Jóvenes adultos  
- `26-40`: Adultos jóvenes  
- `41-60`: Adultos  
- `61+`: Adultos mayores  


In [ ]:
df.columns


La lista confirma que las transformaciones han añadido columnas como `time_of_day` y `diff_time`, mientras se mantienen las columnas clave para el análisis.
Lista de 22 columnas, incluyendo `ve_total`, `fatals`, `peds`, `year`, `monthname`, `day_weekname`, `func_sysname`, `age`, `diff_time`, `time_of_day`.


In [ ]:
column_rename_map = {
    've_total': 'number_of_vehicles_involved',
    'fatals': 'number_of_fatalities',
    'peds': 'pedestrians_involved',
    'year': 'year',
    'monthname': 'month',
    'day_weekname': 'day_of_week',
    'rur_urbname': 'urban_rural',
    'func_sysname': 'road_type',
    'weathername': 'weather_conditions',
    'lgt_condname': 'visibility_level',
    'harm_evname': 'accident_cause',
    'age': 'driver_age_group',
    'sex': 'driver_gender',
    'alc_res': 'driver_alcohol_level',
    'diff_time': 'emergency_response_time',
    'time': 'time_of_day'
}

df.rename(columns=column_rename_map, inplace=True)


El renombramiento hace que los nombres de las columnas sean más descriptivos y consistentes con los datos de que provienen de Postgres y para su posterior modelo dimensional, mejorando la claridad para análisis posteriores

In [ ]:
df.columns

Lista de 22 columnas, incluyendo number_of_vehicles_involved, number_of_fatalities, driver_age_group, emergency_response_time, time_of_day

La lista confirma que las columnas se renombraron correctamente.

In [ ]:
df = df.drop(columns=['hour','minute', 'arr_hour', 'arr_min', 'arr_time'])


Esta celda elimina columnas temporales (`hour`, `minute`, `arr_hour`, `arr_min`, `arr_time`) que ya no son necesarias tras crear `time_of_day` y `emergency_response_time`.

La eliminación reduce la dimensionalidad del DataFrame, manteniendo solo las columnas relevantes para el modelo dimensional.


In [ ]:
df.columns



Lista de 16 columnas: `number_of_vehicles_involved`, `number_of_fatalities`, `pedestrians_involved`, `year`, `month`, `day_of_week`, `urban_rural`, `road_type`, `weather_conditions`, `visibility_level`, `accident_cause`, `driver_age_group`, `driver_gender`, `driver_alcohol_level`, `time_of_day`, `emergency_response_time`.

El DataFrame final está listo para el modelado dimensional, con columnas transformadas y nombres descriptivos.
